### DISCUSSION TITLE
Strange typescript issue

### DISCUSSION THREAD

**Discussion by bellostom** [mobxjs/mobx#3217] [General] [answered]
Hi

I get the attached error from typescript when using observe. However, it works correctly in the browser, there is no error.
Any ideas why? based on the error, it seems that I am not using it correctly, although I have been using observe this way for a long time is pure JS projects.
I am now porting to TS and run into this.
Thanks for any feedback

![ts_error](https://user-images.githubusercontent.com/1138182/145408221-687d0462-a63c-40c4-918f-8987c6767b75.png)

**Comment by kubk** [comment, marked as answer]
Hi @bellostom 
In TypeScript `keyof Type` skips private/protected fields:

```typescript
class Test {
  public a = 1;
  private b = 1;
}

type Props = keyof Test; // Only 'a', not 'a' | 'b'
```

As a workaround you can change `protected` to `public`. As a solution I'd suggest to replace `observe` with `reaction` or `autorun` as recommended in docs: https://mobx.js.org/intercept-and-observe.html#intercept--observe-

Note for Mobx maintainers: The issue can be solved by introducing `AdditionalKeys` on `observe` the same way it was done for `makeObservable`/`makeAutoObservable`:
- https://github.com/mobxjs/mobx/blob/11740048ad1c53c421d045108485a89aafb04ebe/packages/mobx/src/api/makeObservable.ts#L26
- https://mobx.js.org/observable-state.html#limitations

But since `observe` is low level utility and it is not recommended to use, I wouldn't add this TS hack to Mobx repo again.

**Comment by bellostom** [comment]
Thanks @kubk for the comprehensive answer.
I switched to reaction
I would never have thought that the visibility was the culprit
